### ISSUE TITLE
[BUG?] Additional code-blocks being checked since v6.0.0a1 leads to errors not detected before which cannot be specifically ignored

### ISSUE THREAD

**Issue by Cielquan** [rstcheck/rstcheck#120]
Just want to provide one more data point. Issues related to https://github.com/myint/rstcheck/issues/113 have indeed been fixed in the `rc`. However, `rstcheck` now seems to check more code blocks than it did before. While this is obviously a good thing it also makes it somewhat unusable for us over at `pylint`.
We have code that produces `SyntaxErrors` in our changelog to show for which code we can warn when we introduce new checks. I looked in the documentation but there doesn't seem to be a way to ignore specific code blocks. I guess that's what 
#103 is about. Other users might run into similar issues I guess. Not sure if this would be a release blocker for you, but implementation of #103 would be much appreciated 😄
If you would like me to test anything, feel free to ping!

Relevant PR: https://github.com/PyCQA/pylint/pull/6631.

_Originally posted by @DanielNoord in https://github.com/rstcheck/rstcheck/issues/104#issuecomment-1133928874_

**Comment by DanielNoord** [comment]
Just want to stress again: it doesn't seem like the new version introduces an actual regression. In fact, it seems to have fixed a bug and checks for syntactically incorrect code better now. However, something in the discovery of code blocks might have changed which is problematic when we can't ignore specific blocks.

For us, implementation of #103 would solve the issue. Rather than reverting the changes to the discovery system.

**Comment by Cielquan** [comment]
Like I wrote in https://github.com/rstcheck/rstcheck/issues/104#issuecomment-1134317830 I want to keep the usability for the users. Therefore I want this issue to be fixed.

I did not plan to revert anything, because it is actually a good thing like you said, but I want to add options to improve and fix this issue.

**Comment by DanielNoord** [comment]
> Like I wrote in [#104 (comment)](https://github.com/rstcheck/rstcheck/issues/104#issuecomment-1134317830) I want to keep the usability for the users. Therefore I want this issue to be fixed.
> 
> I did not plan to revert anything, because it is actually a good thing like you said, but I want to add options to improve and fix this issue.

Yeah, I thought so but because of the issue titles I was worried you were going to revert the good work you had already done 😄 

Let me know if I can help with anything else!

**Comment by Cielquan** [comment]
The issue is reproducible with v6.0.0a1.

**Comment by Cielquan** [comment]
# Investigations

I use the pylint repository for testing.

## Status-quo

When I run rstcheck via pre-commit, like it is done for pylint, for v5 like currently specified in the repo I get no linting issues.

```yaml
...
  - repo: https://github.com/myint/rstcheck
    rev: "v5.0.0"
    hooks:
      - id: rstcheck
        args: ["--ignore-roles=func,class,mod", "--report=warning"]
        types: [text] # necessary to include ChangeLog file
        files: ^(ChangeLog|doc/(.*/)*.*\.rst)
...
```

```shell
$ pre-commit run rstcheck --all-files
rstcheck.................................................................Passed
```

If I update the config to use v6 and migrate the config to work with v6 alpha1 I get 6 liniting issues.

``` yaml
  - repo: https://github.com/myint/rstcheck
    rev: "v6.0.0a1"
    hooks:
      - id: rstcheck
        args: ["--ignore-roles=func,class,mod", "--report-level=warning"]
        types: [text] # necessary to include ChangeLog file
        files: ^(ChangeLog|doc/(.*/)*.*\.rst)
        additional_dependencies: [sphinx]
```

```shell
$ pre-commit run rstcheck --all-files
rstcheck.................................................................Failed
- hook id: rstcheck
- exit code: 1

doc/whatsnew/2.13.rst:543: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.9.rst:46: (ERROR/3) (python) invalid syntax
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
doc/whatsnew/1.8.rst:60: (ERROR/3) (python) positional argument follows keyword argument
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
Success! No issues detected.
doc/whatsnew/1.7.rst:37: (ERROR/3) (python) expected an indented block
doc/whatsnew/1.7.rst:592: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.7.rst:673: (ERROR/3) (python) expected an indented block
Success! No issues detected.
Success! No issues detected.
```

**Note:**

The `Success! No issues detected.` messages in between are there because pre-commit by default runs tools in parallel. So there were 17 parallel runs.
If I set `require_serial: true` I interestingly get the warning flood from #113, which I did not get before.
When I then update the version to v6.0.0rc1 the warnings disappear and only the 6 errors are printed.
I think I will update rstcheck's hooks config to set `require_serial: true`, because rstcheck itself runs in parallel too,
so there is an overhead for double setup of multiprocessing (I think pre-commit uses multiprocessing too, but I am not sure -- did not check).
The users won't need to do anything, except update to next release.


## Manually running rstcheck

I have two clones of rstcheck on my maschine. One in dir `rstcheck` checked out at master and one in dir `rstcheck_2` checked out at the v5.0.0 tag.

In the pylint repo I set up two venvs `.venv_master` and `.venv_old_tags`:

For `.venv_master` I ran
```shell
$ python3 -m venv .venv_master
$ . .venv_master/bin/activate
$ pip install -U pip
$ pip install sphinx
$ pip install -e ../rstcheck
```

For `.venv_old_tags` I ran
```shell
$ python3 -m venv .venv_old_tags
$ . .venv_old_tags/bin/activate
$ pip install -U pip
$ pip install -e ../rstcheck_2
```

Then I ran rstcheck manually in both venvs:
`.venv_master`:
```shell
$ rstcheck --ignore-roles=func,class,mod --report-level=warning ChangeLog doc -r
doc/whatsnew/2.13.rst:543: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.9.rst:46: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.8.rst:60: (ERROR/3) (python) positional argument follows keyword argument
doc/whatsnew/1.7.rst:37: (ERROR/3) (python) expected an indented block
doc/whatsnew/1.7.rst:592: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.7.rst:673: (ERROR/3) (python) expected an indented block
```

`.venv_old_tags`:
```shell
$ rstcheck --ignore-roles=func,class,mod --report=warning ChangeLog doc -r

```

We can see the mentioned outputs.

### Further testing in `.venv_old_tags`

Like described in https://github.com/rstcheck/rstcheck/issues/109#issuecomment-1129060804 v6 removed the hardcoded default values for sphinx roles and directives.
Because of those defaults pylint was not configured to use the acutal sphinx roles/directives by providing rstcheck with an installed sphinx instance.

When I now install sphinx for v5 and run again, I get:
```shell
$ pip install sphinx
$ rstcheck --ignore-roles=func,class,mod --report=warning ChangeLog doc -r
ChangeLog:1085: (ERROR/3) (python) invalid syntax
doc/whatsnew/2.13.rst:543: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.9.rst:46: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.8.rst:60: (ERROR/3) (python) positional argument follows keyword argument
doc/whatsnew/1.7.rst:37: (ERROR/3) (python) expected an indented block
doc/whatsnew/1.7.rst:592: (ERROR/3) (python) invalid syntax
doc/whatsnew/1.7.rst:673: (ERROR/3) (python) expected an indented block
```

So the issue was always there, but hidden through the defauls I guess.

## Conclusion

This issue is acutally neither a bug nor something new at all. It seems the issue was just hiding behind the sphinx default values for roles and directives to ignore.

The strange thing however is I cannot explain this behavior with the underlying code of v5.

We have the [`register_code_directive`](https://github.com/rstcheck/rstcheck/blob/v5.0.0/rstcheck/__init__.py#L143) function which registers a custom diretive class for `code-block` but only if sphinx is **not** installed.

On the other hand the [`ignore_sphinx`](https://github.com/rstcheck/rstcheck/blob/v5.0.0/rstcheck/__init__.py#L475) function which registers the to ignore sphinx roles and directives has no hardcoded value for `code-block`.

So when sphinx is **not** installed, the custom `code-clock` directive gets registered and definitely no ignore for `code-block` gets registered. So I would guess the linting errors should be found.

And the other hand with sphinx installed the custom `code-block` directive does not get registered and in the loaded directives from sphinx itself is no `code-block` directive.
I know from earlier views into the docutils code that it maps `code-block` to `code` by default.

So the linitng issues should actually be printed in any case to my understanding.


## TL;DR

The issue was hidden until v6 because of some kind of *valid misconfiguration* for v5 and before, so to say.


---

@DanielNoord 

I will close this issue, because there is nothing to fix and the behavior is somewhat actually the same as v5 and before.

I will also release v6 as planned before.
But I am starting work on #103 now and want to release it with v6.1 soon. I would ping you in #103 if I have something you can test if it fixes your issue 😄

**Comment by DanielNoord** [comment]
Thanks for the thorough investigation @Cielquan! That is indeed quite weird...

We'll eagerly await #103 then 😄
